File: engine/core/src/subjects.ts
export const Subjects = {
  createEntity: "engine.world.createEntity",
  deleteEntity: "engine.world.deleteEntity",
  hasEntity: "engine.world.hasEntity",
  listEntities: "engine.world.listEntities",
  addComponent: "engine.entity.addComponent",
  removeComponent: "engine.entity.removeComponent",
  hasComponent: "engine.entity.hasComponent",
  getComponents: "engine.entity.getComponents",
  registerComponent: "engine.component.register",
  startWorker: "engine.worker.start",
  stopWorker: "engine.worker.stop",
} as const;

export const WorkerSubjects = {
  getProperty: (componentId: string, entityId: string, property: string) =>
    `engine.worker.${componentId}.${entityId}.property.${property}.get`,
  setProperty: (componentId: string, entityId: string, property: string) =>
    `engine.worker.${componentId}.${entityId}.property.${property}.set`,
  callMethod: (componentId: string, entityId: string, method: string) =>
    `engine.worker.${componentId}.${entityId}.method.${method}`,
} as const;

File: engine/core/Subjects.cs
namespace Engine.Core;

public static class Subjects
{
    public const string CreateEntity = "engine.world.createEntity";
    public const string DeleteEntity = "engine.world.deleteEntity";
    public const string HasEntity = "engine.world.hasEntity";
    public const string ListEntities = "engine.world.listEntities";

    public const string AddComponent = "engine.entity.addComponent";
    public const string RemoveComponent = "engine.entity.removeComponent";
    public const string HasComponent = "engine.entity.hasComponent";
    public const string GetComponents = "engine.entity.getComponents";
    public const string QueryEntity = "engine.entity.query";

    public const string RegisterComponent = "engine.component.register";
    public const string ListComponents = "engine.component.list";

    public const string StartWorker = "engine.worker.start";
    public const string StopWorker = "engine.worker.stop";
}

public static class WorkerSubjects
{
    public static string CallMethod(string componentId, string entityId, string method) =>
        $"engine.worker.{componentId}.{entityId}.method.{method}";
}
